crates/network/src/any/mod.rs

```diff
@@ -46,11 +59,537 @@ impl From<TxType> for AnyTxType {
+#[derive(Debug, Clone, PartialEq, Eq, Default)]
+#[allow(unnameable_types)]
+pub struct DeserMemo {
+    input: OnceLock<Bytes>,
+    access_list: OnceLock<AccessList>,
+    blob_versioned_hashes: OnceLock<Vec<B256>>,
+    authorization_list: OnceLock<Vec<SignedAuthorization>>,
+}
+
+/// Transaction envelope for a catch-all network.
+#[derive(Debug, Clone, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[serde(untagged)]
+#[doc(alias = "AnyTransactionEnvelope")]
+pub enum AnyTxEnvelope {
+    /// An Ethereum transaction.
+    Ethereum(TxEnvelope),
+    /// A transaction with unknown type.
+    Other {
+        /// Transaction type.
+        #[serde(rename = "type")]
+        ty: AnyTxType,
+        /// Additional fields.
+        fields: std::collections::BTreeMap<String, serde_json::Value>,
+
+        /// Memoization for deserialization.
+        #[serde(skip, default)]
+        memo: DeserMemo,
+    },
+}
+
+impl AnyTxEnvelope {
+    /// Select a field by key and attempt to deserialize it.
+    ///
+    /// This method will return `None` if the key is not present in the fields,
+    /// or if the transaction is already fully deserialized (i.e. it is an
+    /// Ethereum [`TxEnvelope`]). Otherwise, it will attempt to deserialize the
+    /// field and return the result wrapped in a `Some`.
+    pub fn deser_by_key<T: serde::de::DeserializeOwned>(
+        &self,
+        key: &str,
+    ) -> Option<serde_json::Result<T>> {
+        let Self::Other { fields, .. } = self else {
+            return None;
+        };
+
+        fields.get(key).cloned().map(serde_json::from_value)
+    }
+}
+
+impl Encodable2718 for AnyTxEnvelope {
+    fn type_flag(&self) -> Option<u8> {
+        match self {
+            Self::Ethereum(t) => t.type_flag(),
+            Self::Other { ty, .. } => Some(ty.into()),
+        }
+    }
+
+    fn encode_2718_len(&self) -> usize {
+        match self {
+            Self::Ethereum(t) => t.encode_2718_len(),
+            Self::Other { .. } => 1,
+        }
+    }
+
+    #[track_caller]
+    fn encode_2718(&self, out: &mut dyn alloy_primitives::bytes::BufMut) {
+        match self {
+            Self::Ethereum(t) => t.encode_2718(out),
+            Self::Other { ty, .. } => {
+                panic!(
+                    "Attempted to encode unknown transaction type: {}. This is not a bug in alloy. To encode or decode unknown transaction types, use a custom Transaction type and a custom Network implementation. See https://docs.rs/alloy-network/latest/alloy_network/ for network documentation.",
+                    ty
+                )
+            }
+        }
+    }
+
+    fn trie_hash(&self) -> B256 {
+        match self {
+            Self::Ethereum(tx) => tx.trie_hash(),
+            Self::Other { fields, .. } => fields
+                .get("hash")
+                .and_then(|v| v.as_str())
+                .and_then(|v| v.parse().ok())
+                .unwrap_or_else(B256::default),
```

Comment:
let's require `Other` variant to deserialize `hash: B256`

Reply:
in 0ba14955